ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/stats/HiveRelMdDistinctRowCount.java

```diff
@@ -61,17 +62,11 @@ private HiveRelMdDistinctRowCount() {
   @Override
   public Double getDistinctRowCount(RelNode rel, RelMetadataQuery mq, ImmutableBitSet groupKey,
       RexNode predicate) {
-    if (rel instanceof HiveTableScan) {
-      return getDistinctRowCount((HiveTableScan) rel, mq, groupKey, predicate);
-    }
-    /*
-     * For now use Calcite' default formulas for propagating NDVs up the Query
-     * Tree.
-     */
-    return super.getDistinctRowCount(rel, mq, groupKey, predicate);
+    return NumberUtil.multiply(mq.getRowCount(rel),
```

Comment:
We don't take groupKey into consideration anymore? What is the reason behind this change?

Reply:
The logic in the super method was:
```
    // REVIEW zfong 4/19/06 - Broadbase code does not take into
    // consideration selectivity of predicates passed in.  Also, they
    // assume the rows are unique even if the table is not
    boolean uniq = RelMdUtil.areColumnsDefinitelyUnique(mq, rel, groupKey);
    if (uniq) {
      return NumberUtil.multiply(mq.getRowCount(rel),
          mq.getSelectivity(rel, predicate));
    }
    return null;
```
I cannot figure out the logic behind it... if columns are definitely unique, return distinct row count, otherwise return null? 
In any case, I thought about this again and I changed the logic. I left implementation as it was (I removed the method because it was just calling to the super method) and I changed slightly the logic in FilterEstimator so it can handle null values. Basically, null is unknown and you just push those predicates in the AND / OR to the very end since you do not know how selective they are. This may happen when we find an operator in the tree that is not recognized yet in RelMdDistinctRowCount (e.g. DruidQuery).